[MXNET-1121] Example to demonstrate the inference workflow using RNN

leleamol opened the pull request:
## Description ##
 Example to demonstrate the inference workflow using RNN and C++ APIs

## Checklist ##
### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [y] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [y] Changes are complete (i.e. I finished coding on this PR)
- [y] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [y] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [y] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
The example demonstrates inference workflow using pre-trained RNN network and C++ APIs.

Roshrini commented:
@mxnet-label-bot Add [Example, pr-awaiting-review]

lupesko commented on cpp-package/example/inference/simple_rnn.cpp:
```diff
@@ -0,0 +1,359 @@
+                      input_sequence_length);
+
+    // Run the forward pass to predict the image.
+    predict.PredictText(input_sequence);
```
Is it not cleaner when Predictor returns the prediction result, and the calling function prints it?
If you were to use a model in production, would you not design it that way?

lupesko commented on cpp-package/example/inference/simple_rnn.cpp:
```diff
@@ -0,0 +1,359 @@
+#include <string>
+#include <vector>
+#include "mxnet-cpp/MxNetCpp.h"
+#include <opencv2/opencv.hpp>
```
opencv?

lupesko commented on cpp-package/example/inference/simple_rnn.cpp:
```diff
@@ -0,0 +1,359 @@
+    Predictor(const std::string& model_json,
+              const std::string& model_params,
+              const std::string& input_dictionary,
+              bool gpu_context_type = false,
```
Maybe bool use_gpu is a simpler name?

lupesko commented on cpp-package/example/inference/simple_rnn.cpp:
```diff
@@ -0,0 +1,359 @@
+              const std::string& input_dictionary,
+              bool gpu_context_type = false,
+              int sequence_length = 35);
+    void PredictText(const std::string& input_sequence);
```
Shouldn't PredictText return a string?

lupesko commented on cpp-package/example/inference/simple_rnn.cpp:
```diff
@@ -0,0 +1,359 @@
+void Predictor::ConverToIndexVector(const std::string& input, std::vector<float> *input_vector) {
+  std::istringstream input_string(input);
+  input_vector->clear();
+  char delimiter = ' ';
```
const?

lupesko commented on cpp-package/example/inference/simple_rnn.cpp:
```diff
@@ -0,0 +1,359 @@
+   * Convert the input string to a vector of indices that represent
+   * the words in the input string.
+   */
+  std::vector<float> array(sequence_length, 0);
```
You call this variable array although it is a vector, it is a bit confusing and very generic name.
How about naming it `indexVector`?

lupesko commented on cpp-package/example/inference/simple_rnn.cpp:
```diff
@@ -0,0 +1,359 @@
+  Shape input_shape(sequence_length, 1);
+  NDArray input_data = NDArray(input_shape, global_ctx, false);
+  input_data.SyncCopyFromCPU(array.data(), input_shape.Size());
+  NDArray::WaitAll();
```
Do we really need all these WaitAll, or is just one enough before the forward pass?

lupesko commented on cpp-package/example/inference/simple_rnn.cpp:
```diff
@@ -0,0 +1,359 @@
+  // Run the forward pass.
+  executor->Forward(false);
+
+  // The output is available in executor->outputs.
```
Can you say something in the comment about why the 4th index specifically? this seems completely "magical"

lupesko commented on cpp-package/example/inference/simple_rnn.cpp:
```diff
@@ -0,0 +1,359 @@
+  arrayout = arrayout.ArgmaxChannel();
+  NDArray::WaitAll();
+
+  std::ostringstream oss;
```
As I mentioned in other comments, I suggest you consider that Predictor returns the string and the caller function prints it.

leleamol replied:
Modified the PredictText method to return the string.

lupesko commented on cpp-package/example/inference/simple_rnn.cpp:
```diff
@@ -0,0 +1,359 @@
+  if (executor) {
+    delete executor;
+  }
+  MXNotifyShutdown();
```
Can you add a comment explaining this?

lupesko commented on cpp-package/example/inference/simple_rnn.cpp:
```diff
@@ -0,0 +1,359 @@
+  MXNotifyShutdown();
+}
+
+void printUsage() {
```
Can you add an example string as well?

lupesko commented on cpp-package/example/inference/simple_rnn.cpp:
```diff
@@ -0,0 +1,359 @@
+    // Run the forward pass to predict the image.
+    predict.PredictText(input_sequence);
+  } catch (std::runtime_error &error) {
+    LG << "Execution failed with ERROR: " << error.what();
```
return non-zero?

leleamol replied:
Fixed.

lupesko commented on cpp-package/example/inference/simple_rnn.cpp:
```diff
@@ -0,0 +1,359 @@
+     * accessible through MXGetLastError() function.
+     */
+    LG << "Execution failed with following MXNet error";
+    LG << MXGetLastError();
```
return non-zero?

lupesko commented:
@leleamol left a few comments. On the PR description, I suggest you remove the bullets/items that are not relevant for this PR (under Essentials and Changes).

leleamol commented on cpp-package/example/inference/README.md:
````diff
@@ -39,3 +39,35 @@ Alternatively, The script [unit_test_inception_inference.sh](<https://github.com
 ```
+
+### [simple_rnn.cpp](<https://github.com/apache/incubator-mxnet/blob/master/cpp-package/example/inference/simple_rnn.cpp>)
+This example demonstrates sequence prediction workflow with pre-trained RNN model using MXNet C++ API. The purpose of this example is to demonstrate how pre-trained RNN model can be loaded and used to generate an output sequence using C++ API.
````
Fixed.

leleamol commented on cpp-package/example/inference/README.md:
```diff
@@ -39,3 +39,35 @@ Alternatively, The script [unit_test_inception_inference.sh](<https://github.com
+### [simple_rnn.cpp](<https://github.com/apache/incubator-mxnet/blob/master/cpp-package/example/inference/simple_rnn.cpp>)
+This example demonstrates sequence prediction workflow with pre-trained RNN model using MXNet C++ API. The purpose of this example is to demonstrate how pre-trained RNN model can be loaded and used to generate an output sequence using C++ API.
+The example performs following tasks.
+- Load the pre-trained RNN model,
```
Fixed.

leleamol commented on cpp-package/example/inference/README.md:
```diff
@@ -39,3 +39,35 @@ Alternatively, The script [unit_test_inception_inference.sh](<https://github.com
+
+### [simple_rnn.cpp](<https://github.com/apache/incubator-mxnet/blob/master/cpp-package/example/inference/simple_rnn.cpp>)
+This example demonstrates sequence prediction workflow with pre-trained RNN model using MXNet C++ API. The purpose of this example is to demonstrate how pre-trained RNN model can be loaded and used to generate an output sequence using C++ API.
+The example performs following tasks.
```
Fixed.

leleamol commented on cpp-package/example/inference/README.md:
```diff
@@ -39,3 +39,35 @@ Alternatively, The script [unit_test_inception_inference.sh](<https://github.com
+- [obama-speaks-0100.params](<https://s3.amazonaws.com/mxnet-cpp/RNN_model/obama-speaks-0100.params>)
+- [obama.dictionary.txt](<https://s3.amazonaws.com/mxnet-cpp/RNN_model/obama.dictionary.txt>) Each line of the dictionary file contains a word and an unique index for that word separated by a space. This dictionary file is created when the RNN model was trained with a particular dataset.
+
+The example downloads the above files while running. The example is intentionally kept specific to a certain model because:
```
Fixed.

leleamol commented on cpp-package/example/inference/README.md:
```diff
@@ -39,3 +39,35 @@ Alternatively, The script [unit_test_inception_inference.sh](<https://github.com
+The example uses a pre-trained RNN model that is trained with the dataset containing speeches given by Obama. The model files can be found here.
+- [obama-speaks-symbol.json](<https://s3.amazonaws.com/mxnet-cpp/RNN_model/obama-speaks-symbol.json>)
+- [obama-speaks-0100.params](<https://s3.amazonaws.com/mxnet-cpp/RNN_model/obama-speaks-0100.params>)
+- [obama.dictionary.txt](<https://s3.amazonaws.com/mxnet-cpp/RNN_model/obama.dictionary.txt>) Each line of the dictionary file contains a word and an unique index for that word separated by a space. This dictionary file is created when the RNN model was trained with a particular dataset.
```
Fixed.

leleamol commented:
@lupesko 
Thanks for the review comments. I have updated the code as per your comments.

Also I have edited the README files with some details about the model as well as instruction to build cpp package before building the examples.

leleamol commented on cpp-package/example/inference/README.md:
```diff
@@ -39,3 +39,35 @@ Alternatively, The script [unit_test_inception_inference.sh](<https://github.com
+The example downloads the above files while running. The example is intentionally kept specific to a certain model because:
+- Unavailability of pre-trained RNN models that can be successfully imported using C++.
+- C++ API currently does not support bucketing. Hence the shape of input data needs to be fixed and known while loading the model.
+**TODO**: The example can later be enhanced once the above limitations are addressed.
```
Done.

leleamol commented:
@sandeep-krishnamurthy

leleamol commented on cpp-package/example/inference/simple_rnn.cpp:
```diff
@@ -0,0 +1,376 @@
+  std::string input_dictionary = "./obama.dictionary.txt";
+  std::string input_sequence = "Good morning. I appreciate the opportunity to speak here";
+
+  int input_sequence_length = 35;
```
We are padding the sentence if users do not give sequence of 35 words.

leleamol commented on cpp-package/example/inference/simple_rnn.cpp:
```diff
@@ -0,0 +1,376 @@
+  ConverToIndexVector(input_text, &index_vector);
+
+  Shape input_shape(sequence_length, 1);
+  NDArray input_data = NDArray(input_shape, Context::cpu(), false);
```
Fixed. We don't need this NDarray actually.

leleamol commented on cpp-package/example/inference/README.md:
````diff
@@ -39,3 +39,48 @@ Alternatively, The script [unit_test_inception_inference.sh](<https://github.com
+
+The example will output the seqence of 35 words as follows:
+```
+[waters elected Amendment Amendment Amendment Amendment retirement maximize maximize maximize acr citi sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio so ...
````
I tried to get better output by changing model hyperparameters but couldn't get it. It would require a good amount of input data processing as well. All these efforts would require dedicated time and out of scope for this example. The example aims towards loading the model and running forward pass. Improving on the model would be a separate task.

leleamol replied:
I am working on implementing the RNN model using C++ API. I can work on improving the accuracy of that model and use it in this example later.

leleamol commented on cpp-package/example/inference/README.md:
```diff
@@ -39,3 +39,48 @@ Alternatively, The script [unit_test_inception_inference.sh](<https://github.com
+
+The example uses a pre-trained RNN model that is trained with the dataset containing speeches given by Obama.
+The model consists of :
+- Embedding Layer with the size of embedding to be 650
```
I can get the pdf version of model (generated using visualiation API). Need some suggestions from @aaronmarkham  to embed them in README.

leleamol commented on cpp-package/example/inference/README.md:
````diff
@@ -39,3 +39,48 @@ Alternatively, The script [unit_test_inception_inference.sh](<https://github.com
+./simple_rnn --input "Good morning. I appreciate the opportunity to speak here"
+```
+
+The example will output the seqence of 35 words as follows:
````
Done.

leleamol commented on cpp-package/example/inference/simple_rnn.cpp:
```diff
@@ -0,0 +1,376 @@
+    throw std::runtime_error("Model parameters does not exist");
+  }
+  LG << "Loading the model parameters from " << model_parameters_file << std::endl;
+  std::map<std::string, NDArray> paramters;
```
Done.

leleamol commented on cpp-package/example/inference/simple_rnn.cpp:
```diff
@@ -0,0 +1,376 @@
+   * The output "softmax_layer_output" has shape [sequence_length, vocab_size]
+   * The vocab_size in this case is equal to the size of dictionary.
+   */
+  int output_index = GetIndexForOutputSymbolName("softmax_layer_output");
```
Done

leleamol commented on cpp-package/example/inference/README.md:
```diff
@@ -39,3 +39,48 @@ Alternatively, The script [unit_test_inception_inference.sh](<https://github.com
+
+The model files can be found here.
+- [obama-speaks-symbol.json](<https://s3.amazonaws.com/mxnet-cpp/RNN_model/obama-speaks-symbol.json>)
+- [obama-speaks-0100.params](<https://s3.amazonaws.com/mxnet-cpp/RNN_model/obama-speaks-0100.params>)
```
The current bucket mxnet-cpp is not a public by default but the contents are made publicly readable. This is similar to mxnet-scala bucket used for scala examples.

leleamol commented:
@sandeep-krishnamurthy as per the comment, I have added these examples to ci_test.sh file that runs in nightly builds.

piyushghai commented on cpp-package/example/inference/simple_rnn.cpp:
```diff
@@ -0,0 +1,372 @@
+
+
+int main(int argc, char** argv) {
+  std::string model_file_json = "./obama-speaks-symbol.json";
```
Should the input file paths be hardcoded here ? 
Can they be passed as parameters to the main method in ```argv``` ?

piyushghai replied:
My bad, saw later that these are passed to the S3 Bucket to download.

IvyBazan commented on cpp-package/example/inference/README.md:
```diff
@@ -39,3 +39,48 @@ Alternatively, The script [unit_test_inception_inference.sh](<https://github.com
+### [simple_rnn.cpp](<https://github.com/apache/incubator-mxnet/blob/master/cpp-package/example/inference/simple_rnn.cpp>)
+This example demonstrates sequence prediction workflow with pre-trained RNN model using MXNet C++ API. The purpose of this example is to demonstrate how a pre-trained RNN model can be loaded and used to generate an output sequence using C++ API.
+The example performs following tasks
+- Load the pre-trained RNN model.
```
```suggestion
- Loads the pre-trained RNN model.
```

IvyBazan commented on cpp-package/example/inference/README.md:
```diff
@@ -39,3 +39,48 @@ Alternatively, The script [unit_test_inception_inference.sh](<https://github.com
+This example demonstrates sequence prediction workflow with pre-trained RNN model using MXNet C++ API. The purpose of this example is to demonstrate how a pre-trained RNN model can be loaded and used to generate an output sequence using C++ API.
+The example performs following tasks
+- Load the pre-trained RNN model.
+- Load the dictionary file that contains word to index mapping.
```
```suggestion
- Loads the dictionary file containing the word to index mapping.
```

IvyBazan commented on cpp-package/example/inference/README.md:
```diff
@@ -39,3 +39,48 @@ Alternatively, The script [unit_test_inception_inference.sh](<https://github.com
+The example performs following tasks
+- Load the pre-trained RNN model.
+- Load the dictionary file that contains word to index mapping.
+- Convert the input string to vector of indices and padded to match the input data length.
```
```suggestion
- Converts the input string to vector of indices that's padded to match the input data length.
```

IvyBazan commented on cpp-package/example/inference/README.md:
```diff
@@ -39,3 +39,48 @@ Alternatively, The script [unit_test_inception_inference.sh](<https://github.com
+- Load the pre-trained RNN model.
+- Load the dictionary file that contains word to index mapping.
+- Convert the input string to vector of indices and padded to match the input data length.
+- Run the forward pass and predict the output string.
```
```suggestion
- Runs the forward pass and predicts the output string.
```

IvyBazan commented on cpp-package/example/inference/README.md:
```diff
@@ -39,3 +39,48 @@ Alternatively, The script [unit_test_inception_inference.sh](<https://github.com
+- Convert the input string to vector of indices and padded to match the input data length.
+- Run the forward pass and predict the output string.
+
+The example uses a pre-trained RNN model that is trained with the dataset containing speeches given by Obama.
```
```suggestion
The example uses a pre-trained RNN model trained with a dataset containing speeches given by Obama.
```

IvyBazan commented on cpp-package/example/inference/README.md:
```diff
@@ -39,3 +39,48 @@ Alternatively, The script [unit_test_inception_inference.sh](<https://github.com
+
+The example uses a pre-trained RNN model that is trained with the dataset containing speeches given by Obama.
+The model consists of :
+- Embedding Layer with the size of embedding to be 650
```
```suggestion
- Embedding Layer with the size of embedding set at 650
```

IvyBazan commented on cpp-package/example/inference/README.md:
````diff
@@ -39,3 +39,48 @@ Alternatively, The script [unit_test_inception_inference.sh](<https://github.com
+[waters elected Amendment Amendment Amendment Amendment retirement maximize maximize maximize acr citi sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio so ...
+```
+
+Alternatively, user can run [unit_test_simple_rnn.sh](<https://github.com/apache/incubator-mxnet/blob/master/cpp-package/example/inference/unit_test_simple_rnn.sh>) script.
````
```suggestion
Alternatively, you can run the [unit_test_simple_rnn.sh](<https://github.com/apache/incubator-mxnet/blob/master/cpp-package/example/inference/unit_test_simple_rnn.sh>) script.
```

IvyBazan commented on cpp-package/example/inference/simple_rnn.cpp:
```diff
@@ -0,0 +1,372 @@
+ * 3. input_dictionary: File containing the word and associated index.
+ * 4. sequence_length: Sequence length for which the RNN was trained.
+ *
+ * The constructor will:
```
```suggestion
 * The constructor:
```

IvyBazan commented on cpp-package/example/inference/simple_rnn.cpp:
```diff
@@ -0,0 +1,372 @@
+ * 4. sequence_length: Sequence length for which the RNN was trained.
+ *
+ * The constructor will:
+ *  1. Load the model and parameter files.
```
```suggestion
 *  1. Loads the model and parameter files.
```

IvyBazan commented on cpp-package/example/inference/simple_rnn.cpp:
```diff
@@ -0,0 +1,372 @@
+
+
+/*
+ * The constructor takes following parameters as input:
```
```suggestion
 * The constructor takes the following parameters as input:
```

IvyBazan commented on cpp-package/example/inference/simple_rnn.cpp:
```diff
@@ -0,0 +1,372 @@
+ *
+ * The constructor will:
+ *  1. Load the model and parameter files.
+ *  2. Load the dictionary file to create index to word and word to index maps.
```
```suggestion
 *  2. Loads the dictionary file to create index to word and word to index maps.
```

IvyBazan commented on cpp-package/example/inference/simple_rnn.cpp:
```diff
@@ -0,0 +1,372 @@
+ * The constructor will:
+ *  1. Load the model and parameter files.
+ *  2. Load the dictionary file to create index to word and word to index maps.
+ *  3. Invoke the SimpleBind to bind the input argument to the model and create an executor.
```
```suggestion
 *  3. Invokes the SimpleBind to bind the input argument to the model and create an executor.
```

IvyBazan commented on cpp-package/example/inference/simple_rnn.cpp:
```diff
@@ -0,0 +1,372 @@
+
+
+/*
+ * The function populates the input vector with indices from dictionary that are
```
```suggestion
 * The function populates the input vector with indices from the dictionary that
```

IvyBazan commented on cpp-package/example/inference/simple_rnn.cpp:
```diff
@@ -0,0 +1,372 @@
+
+/*
+ * The function populates the input vector with indices from dictionary that are
+ * corresponding to the words in the input string.
```
```suggestion
 * correspond to the words in the input string.
```

leleamol commented on cpp-package/example/inference/README.md:
```diff
@@ -55,6 +55,7 @@ The model consists of :
 - FullyConnected Layer
 - SoftmaxOutput
 The model was trained for 100 epochs.
+The visual representation of the model is [here](<https://github.com/apache/incubator-mxnet/blob/master/cpp-package/example/inference/obama-speaks.pdf>).
```
Thanks! I have added it now.

leleamol commented:
@mxnet-jenkins The unix-gpu build has failed because the cpp-package:GPU task failed with following error:

> `cp: cannot stat '../../build/cpp-package/example/inception_inference': No such file or directory`

However, I verified that inception_inference is getting built and exists in ./build/cpp-package/example/ folder. It is also getting unpacked as correctly. 
Unpacked lib/libmxnet.so, lib/libmxnet.a, 3rdparty/dmlc-core/libdmlc.a, 3rdparty/tvm/nnvm/lib/libnnvm.a, 3rdparty/ps-lite/build/libps.a, deps/lib/libprotobuf-lite.a, deps/lib/libzmq.a, build/cpp-package/example/* from gpu

> + echo lib/libmxnet.so, lib/libmxnet.a, 3rdparty/dmlc-core/libdmlc.a, 3rdparty/tvm/nnvm/lib/libnnvm.a, 3rdparty/ps-lite/build/libps.a, deps/lib/libprotobuf-lite.a, deps/lib/libzmq.a, build/cpp-package/example/alexnet build/cpp-package/example/alexnet.d build/cpp-package/example/charRNN build/cpp-package/example/charRNN.d build/cpp-package/example/googlenet build/cpp-package/example/googlenet.d build/cpp-package/example/inception_bn build/cpp-package/example/inception_bn.d build/cpp-package/example/inception_inference build/cpp-package/example/inception_inference.d build/cpp-package/example/lenet build/cpp-package/example/lenet.d build/cpp-package/example/lenet_with_mxdataiter build/cpp-package/example/lenet_with_mxdataiter.d build/cpp-package/example/mlp build/cpp-package/example/mlp.d build/cpp-package/example/mlp_cpu build/cpp-package/example/mlp_cpu.d build/cpp-package/example/mlp_csv build/cpp-package/example/mlp_csv.d build/cpp-package/example/mlp_gpu build/cpp-package/example/mlp_gpu.d build/cpp-package/example/resnet build/cpp-package/example/resnet.d build/cpp-package/example/simple_rnn build/cpp-package/example/simple_rnn.d build/cpp-package/example/test_optimizer build/cpp-package/example/test_optimizer.d build/cpp-package/example/test_score build/cpp-package/example/test_score.d

Can somebody help me debug this?

leleamol commented:
@lupesko @sandeep-krishnamurthy @szha 

Since the model used for sequence generation was not giving the good results. I have updated the example.
The example now uses a sentiment analysis model that is built and exported using GluonAPI.
The model is based on the [Sentiment Analysis Gluon Tutorial](http://gluon-nlp.mxnet.io/examples/sentiment_analysis/sentiment_analysis.html#)

I will be updating the ReadMe files in the next commit.

leleamol commented:
@lupesko @sandeep-krishnamurthy @szha @ddavydenko 
I have updated the PR with latest readme files, unit tests. The unit tests are also integrated into nightly CI tests. All CI tests have passed. Can you please review the PR?

sandeep-krishnamurthy commented on cpp-package/example/inference/README.md:
```diff
@@ -39,3 +39,50 @@ Alternatively, The script [unit_test_inception_inference.sh](<https://github.com
+Usage:
+sentiment_analysis_rnn
+--input Input movie review line.e.g. "This movie is the best"
+[--max_num_words]  The number of words in the sentence to be considered for sentiment analysis. Default is 5
```
why 5? I think a usual sentence will be longer than this.

sandeep-krishnamurthy replied:
- I suggest adding more examples of different lengths so it will be clear for users.
- Can we make it clear that if input is less than the length of max_num_words it gets padded with 0s and if it is longer it gets trimmed.

leleamol replied:
Thanks @sandeep-krishnamurthy  for review.
I have increased the length to 10. This can also be specified as a command line parameters.
I have included examples of various lengths in README file.
The "Help" message is updated.

sandeep-krishnamurthy commented on cpp-package/example/inference/sentiment_analysis_rnn.cpp:
```diff
@@ -0,0 +1,397 @@
+   * NDArray. We need to find the index in that vector that
+   * corresponds to the output symbol "sentimentnet0_hybridsequential0_dense0_fwd_output".
+   */
+  const std::string output_symbol_name = "sentimentnet0_hybridsequential0_dense0_fwd_output";
```
This will cause problem for users trying to copy this code for their usecase or using a different model files. Can we think of simplifying this or making it more clear?

leleamol replied:
The Forward() function returns a vector of NDArray each corresponding to output symbols.
Without knowing the name of symbol that is going to output the score, we can not find the exact index in that vector.
As long as users continue to use the same model, the symbol name won't change.
If users change the model, they would need to know the name of output symbol.

stephenrawls replied:
I think your model only has one output, right?  [I verified by downloading model and confirming it only has one output]

So the ListOutputs() function only return a single element. You don't need to provide an explicit name, you can just use the one and single output element.

(It's true that if someone else copies this, and *their* model has multiple output elements, they would need to change the logic.)

leleamol commented:
@szha 
The example accepts input line and a parameter "max_num_words".  The loaded symbol is bound with "max_num_words" which as default value of 10.

The input line is padded or trimmed as necessary to match the "max_num_words".
Since C++ does not have bucketing module yet, the example binds the symbol to a fixed length of input at a time.
The bucketing can be implemented in the example. However, I would prefer to implement it in C++ binding and then modify this example in later versions.

Isa-rentacs commented on cpp-package/example/inference/unit_test_sentiment_analysis_rnn.sh:
```diff
@@ -0,0 +1,36 @@
+else
+    LD_LIBRARY_PATH=${LD_LIBRARY_PATH}:../../../lib ./${EXE_NAME} --input "This movie has the great story and best acting" 2&> ${EXE_NAME}.log
+fi
+result=`grep "The sentiment score between 0 and 1.*\=" sentiment_analysis_rnn.log | cut -d '=' -f2`
```
Replace ```sentiment_analysis_rnn``` with ```EXE_NAME``` ?

leleamol replied:
Done

Isa-rentacs commented on cpp-package/example/inference/unit_test_sentiment_analysis_rnn.sh:
```diff
@@ -0,0 +1,36 @@
+set -e # exit on the first error
+export EXE_NAME=sentiment_analysis_rnn
+
+# Running the example with dog image.
```
sentiment analysis :)

Isa-rentacs commented on cpp-package/example/inference/README.md:
```diff
@@ -39,3 +39,72 @@ Alternatively, The script [unit_test_inception_inference.sh](<https://github.com
+Usage:
+sentiment_analysis_rnn
+--input Input movie review line.e.g. "This movie is the best." The input is trimmed or padded with 0s to match the max_num_words.
+[--max_num_words]  The number of words in the sentence to be considered for sentiment analysis. Default is 10
```
Is there any upper bound for ```max_num_words```?

Isa-rentacs commented:
@leleamol 
Regarding variable length inputs, isn't it possible to update ```args_map["data0"]``` depending on the length of the input sentence?

leleamol commented on cpp-package/example/inference/sentiment_analysis_rnn.cpp:
```diff
@@ -0,0 +1,398 @@
+   * the words in the input string.
+   */
+  std::vector<float> index_vector(num_words, GetIndexForWord("<eos>"));
+  int num_words = ConverToIndexVector(input_text, &index_vector);
```
@Isa-rentacs 
I have updated the example to handle variable length input. I have addressed your review comments in the updated example.
The model expects 2 inputs viz. "data0" and "data1". The "data0" represents the review line for which the score is generated and "data1" represents the number of words in the sentence that are considered for the prediction.  For both the inputs the model expects "NDArray" input.
The sample code line in the [Gluon Tutorial here](http://gluon-nlp.mxnet.io/examples/sentiment_analysis/sentiment_analysis.html#) in python also uses and NDArray of dimension 1.

Regarding the names "data0" and "data1", the model is developed and trained in python by following the above tutorial.  As per the documentation [here](http://mxnet.incubator.apache.org/api/python/gluon/gluon.html#mxnet.gluon.HybridBlock.export) when Gluon model is exported it automatically names the inputs as data0, data1, etc.

Isa-rentacs replied:
Thank you for your comment, I understood about the naming.

leleamol commented:
@szha @Isa-rentacs @ddavydenko @lupesko @sandeep-krishnamurthy 

I have updated the example that can handle the variable length input. The readme and unit test files are also updated accordingly.

szha commented on cpp-package/example/inference/sentiment_analysis_rnn.cpp:
```diff
@@ -0,0 +1,464 @@
+    Executor* executor = net.SimpleBind(global_ctx, args_map, std::map<std::string, NDArray>(),
+                                std::map<std::string, OpReqType>(), aux_map);
+    executor_buckets[bucket] = executor;
+  }
```
This won't share the memory across executors, but will allocate excessive space from which only `1 / bucket_keys.Size()` is used at any time.

stephenrawls replied:
Do you have to make sure that the master executor is created with the largest bucket size? If not, wouldn't the other executors need to allocate additional memory since the master executor only allocated enough for the smaller shape?

It looks like you are looping in whatever order bucket_keys is, and it looks like you are starting from smallest and going to largest bucket size.

(@szha -- you might know better than me. I know in the executor Reshape() call this is true. I would assume it must be true in the BindEX call that is being made here, but not 100% sure)

leleamol replied:
The shared executors that I am creating are for the keys smaller to largest.
However, I noticed that the results do not change or example does not crash even if I reverse the order.
I had single stepped through the code, the GraphExecutor::InitDataEntryMemory() uses the shared memory from the given executor.

stephenrawls replied:
@leleamol  -- the question I had is not weather it will effect the correctness or cause your program to crash. I am wondering weather the current order ends up allocating more memory than you need.

I don't know for sure, but I suspect if you allocate the largest executor first, then all the rest of the executors don't have to allocate extra memory and they just share memory from the largest executor.

However, I would assume that if the smallest executor is the "master" executor, then all other executors would have to allocate extra memory because the master executor did not allocate enough to satisfy any of the other executors.

Can you re-run your test w/ changing the order, and look closely at total gpu memory allocated?

leleamol replied:
@stephenrawls 
I tried running it on CPU and I see an improvement in the memory consumption when the order is reveresed. That is creating the executor with largest bucket first.
I have updated the PR with that change.

stephenrawls replied:
Great, thanks for confirming.

stephenrawls replied:
Just looked at your fix though.

Instead of changing the order of your default array, you should probably just have a check to see what the max element of your array is, and allocate that first.

Otherwise you have a hidden assumption that the bucket sizes should be in a specific order, and it is easy for users not to realize that.

leleamol replied:
@stephenrawls 
I agree, it will imply that the order is important.
Fixed the code in latest update.

stephenrawls replied:
Awesome, it looks ready to go to me. Thanks.  👍

szha commented on cpp-package/example/inference/sentiment_analysis_rnn.cpp:
```diff
@@ -0,0 +1,495 @@
+ * If the next larger bucket does not exist, function looks for the previous bucket key
+ * that is lesser than given num_words.
+ */
+int Predictor::GetClosestBucketKey(int num_words) {
```
it seems that you can simply use lower_bound and return the largest key if not found.

leleamol replied:
If we use lower_bound, we will always find a bucket that is smaller than the input word length (unless the input word length is smaller than the smallest bucket size). In that case, we will trim the input line and may lose the information.  Therefore, I chose to find the next largest bucket so that we will be able to consider all the words in input sentence.

szha replied:
> If we use lower_bound, we will always find a bucket that is smaller than the input word length

I'm not sure if this is true. From cpprefenrece on [std::map::lower_bound](https://en.cppreference.com/w/cpp/container/map/lower_bound):

> Returns an iterator pointing to the first element that is not less than key.

In your case the `key` is input word length.

leleamol replied:
@szha I agree, with lower_bounds we will get matching or a smaller bucket key.  My concern is that, we will always find such bucket and we will trim the input sequence. In that process, we would lose some of the information. Will it be acceptable?

szha replied:
Unless I read the document wrong, it won't get you smaller key. `lower_bound` returns the first element (i.e. bucket key) that's no less than (i.e. larger or equal to) the key (i.e. input sequence length).

szha replied:
For input sequence that's larger than the largest key, clipping is necessary.

leleamol replied:
@szha  Fixed the logic in the latest update. Getting the lower bound is sufficient and accurate.
Thanks!

szha commented on cpp-package/example/inference/sentiment_analysis_rnn.cpp:
```diff
@@ -0,0 +1,495 @@
+/*
+ * The function downloads the model files from s3 bucket.
+ */
+void Download_files(const std::vector<std::string> model_files) {
```
nit: use a consistent naming convention.

leleamol replied:
Done.

szha commented on cpp-package/example/inference/unit_test_sentiment_analysis_rnn.sh:
```diff
@@ -0,0 +1,36 @@
+    LD_LIBRARY_PATH=${LD_LIBRARY_PATH}:../../../lib ./${EXE_NAME}  --input "This movie is the best." 2&> ${EXE_NAME}.log
+fi
+result=`grep "The sentiment score between 0 and 1.*\=" ${EXE_NAME}.log | cut -d '=' -f2`
+if [ $result ==  "0.964498" ];
```
this check seems quite exact... maybe hard to reproduce for the purpose of testing this example. use a range?

leleamol replied:
Done.

sandeep-krishnamurthy commented:
Thanks Amol and all reviewers for this contribution.
Merging the changes as all reviewer comments are addressed and tests have passed.

leleamol commented:
@mxnet-label-bot remove [pr-awaiting-review]
